src/seraphis_wallet/key_container.cpp

```diff
@@ -0,0 +1,223 @@
+
+    if (!read_encrypted_file(path, key, ser_key_container))
+        return {};
+
+    const KeyContainer key_container = ser_container_to_usable(ser_key_container, key);
+
+    if (!jamtis_keys_valid(key_container.m_keys))
+    {
+        return boost::none;
+    }
+
+    return key_container;
+}
+//-------------------------------------------------------------------------------------------------------------------
+bool KeyContainer::encrypt(const crypto::chacha_key &key)
+{
+    if (m_encrypted)
+        return false;
+
+    m_encryption_iv = crypto::rand<crypto::chacha_iv>();
+
+    m_keys.encrypt(key, m_encryption_iv);
+    return true;
+}
+//-------------------------------------------------------------------------------------------------------------------
+bool KeyContainer::decrypt(const crypto::chacha_key &key)
+{
+    if (!m_encrypted)
+        return false;
+
+    m_keys.decrypt(key, m_encryption_iv);
+    return true;
+}
+//-------------------------------------------------------------------------------------------------------------------
+const jamtis_keys &KeyContainer::get_keys()
+{
+    return m_keys;
+}
+//-------------------------------------------------------------------------------------------------------------------
+void KeyContainer::set_keys(jamtis_keys &&keys, const crypto::chacha_key &key)
+{
+    m_keys = keys;
+    m_encrypted = false;
+
+    encrypt(key);
+}
+//-------------------------------------------------------------------------------------------------------------------
+bool KeyContainer::write_all(const std::string &path, const crypto::chacha_key &key)
+{
+    ser_jamtis_keys ser_keys = {
+        .k_m = m_keys.k_m,
+        .k_vb = m_keys.k_vb,
+        .xk_ua = m_keys.xk_ua,
+        .xk_fr = m_keys.xk_fr,
+        .s_ga = m_keys.s_ga,
+        .s_ct = m_keys.s_ct,
+        .K_1_base = m_keys.K_1_base,
+        .xK_ua = m_keys.xK_ua,
+        .xK_fr = m_keys.xK_fr,
+    };
+
+    return write_encrypted_file(path, key, ser_keys);
+}
+//-------------------------------------------------------------------------------------------------------------------
+bool KeyContainer::write_view_only(const std::string &path, const crypto::chacha_key &key)
+{
+    if (m_encrypted)
+        return false;
+
+    ser_jamtis_keys view_only_keys = {
+        .k_m = {},
+        .k_vb = {},
+        .xk_ua = {},
+        .xk_fr = m_keys.xk_fr,
+        .s_ga = {},
+        .s_ct = {},
+        .K_1_base = {},
+        .xK_ua = m_keys.xK_ua,
+        .xK_fr = m_keys.xK_fr,
+    };
+
+    // const ser_KeyContainer ser = {
+    //     .keys = ser_jamtis_keys{m_keys},
+    //     .encrypted = m_encrypted,
+    //     .encryption_iv = m_encryption_iv,
+    // };
+
+    return write_encrypted_file(path, key, view_only_keys);
+}
+//-------------------------------------------------------------------------------------------------------------------
+bool KeyContainer::write_view_balance(const std::string &path, const crypto::chacha_key &key)
```

Comment:
I think it's a quite difficult question to decide how to implement different wallet types, like full wallet, read-only wallet, address-generation wallet, multisig wallet, whatever. I am not sure at all whether finally the way to go will be to have a set of different read and write methods in the key container class here. Maybe leave those away for the time being? But seems to me I should not decide this alone, eager to read what other people are thinking.

Reply:
I think these are convenient for the tiers that make sense to share around